src/Features/Core/Portable/UseAutoProperty/UseAutoPropertyFixAllProvider.cs

```diff
@@ -28,29 +29,61 @@ private sealed class UseAutoPropertyFixAllProvider(TProvider provider) : FixAllP
-                foreach (var (_, diagnostics) in documentToDiagnostics)
+            return await ProducerConsumer<(DocumentId documentId, SyntaxNode newRoot)>.RunParallelAsync(
+                contexts,
+                produceItems: async static (currentContext, callback, args, cancellationToken) =>
                 {
-                    foreach (var diagnostic in diagnostics)
+                    // Within a single context (a project) get all diagnostics, and then handle each diagnostic, one at
+                    // a time, to get the final state of the project.
+                    var (originalContext, provider) = args;
+
+                    // Complete a progress item as we finish each project.
+                    using var _ = originalContext.Progress.ItemCompletedScope();
+
+                    var originalSolution = originalContext.Solution;
+                    var currentSolution = originalSolution;
+
+                    var documentToDiagnostics = await FixAllContextHelper.GetDocumentDiagnosticsToFixAsync(currentContext).ConfigureAwait(false);
+                    foreach (var (_, diagnostics) in documentToDiagnostics)
                     {
-                        currentSolution = await provider.ProcessResultAsync(
-                            originalSolution, currentSolution, diagnostic, cancellationToken).ConfigureAwait(false);
+                        foreach (var diagnostic in diagnostics)
+                        {
+                            currentSolution = await provider.ProcessResultAsync(
+                                originalSolution, currentSolution, diagnostic, cancellationToken).ConfigureAwait(false);
+                        }
                     }
-                }
-            }
 
-            return currentSolution;
+                    // After we finish this context, report the changed documents to the consumeItems callback to process.
+                    // This also lets us release all the forked solution info we created above.
+                    foreach (var changedDocumentId in originalSolution.GetChangedDocuments(currentSolution))
+                    {
+                        var changedDocument = currentSolution.GetRequiredDocument(changedDocumentId);
+                        var changedRoot = await changedDocument.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                        callback((changedDocumentId, changedRoot));
+                    }
+                },
+                consumeItems: async static (documentsIdsAndNewRoots, args, cancellationToken) =>
+                {
+                    // Now take all the changed documents and produce the final solution with all of them combined.
+                    var currentSolution = args.originalContext.Solution;
+                    await foreach (var (documentId, newRoot) in documentsIdsAndNewRoots)
+                        currentSolution = currentSolution.WithDocumentSyntaxRoot(documentId, newRoot);
```

Comment:
>WithDocumentSyntaxRoot [](http://example.com/codeflow?start=58&length=22)

WithDocumentSyntaxRoots?

Also, would it work to just have the consume code collect into an array and then once the Producer/Consumer is done, have it make a single call to WithDocumentSyntaxRoots?

Reply:
Or maybe the 2nd part of that comment won't work, forgot there was both an ImmutableArray and IAsyncEnumerable version of these calls.

Reply:
OK, good enough. FYI, I was thinking that the code would be simpler by using the form that didn't pass in a consumeItems parameter, and just did the WithDocumentSyntaxRoots outside the RunParallelAsync call (or maybe I'm confused)

Reply:
i forgot about those.  That is simpler.  Thanks!